### ISSUE TITLE
ipfshttpclient version is incompatible with python version 3.7

### ISSUE THREAD

**Issue by unparalleled-js** [ethereum/web3.py#2378]
* Version:5.27.0
* python 3.7 and 3.7.1

### What was wrong?

In the `setup.py` file, there is `python_requires=">=3.7,<3.11",` but there is also `ipfshttpclient==0.8.0a2"`.
But, when looking at `ipfshttpclient`, it does not support python 3.7 and 3.7.1: https://github.com/ipfs-shipyard/py-ipfs-http-client/blob/master/pyproject.toml#L15-L23

Thus, web3.py 5.27.0 does not support python 3.7 and 3.7.1

This is causing downstream dependencies such as https://github.com/software-mansion/starknet.py to be stuck on older versions of web3.py.

### How can it be fixed?

* Officially stop supporting python 3.7 and 3.7.1 e.g. `python_requires=">=3.7.2,<3.11",`

---
**Note:** We prefer to use issues to track our work. If you think you've encountered a bug in web3py or 
have a feature request, you're in the right place. If you have implementation or usage questions, 
please refer to our [documentation](https://web3py.readthedocs.io/en/latest/) and/or join the conversation 
on [discord](https://discord.gg/GHryRvPB84).

**Comment by unparalleled-js** [comment]
My experience is that when I install or use a python version, it just about always is the latest minor in that range, like `3.7.12` or something, so no no one probably experiences this bug... But, tools like `poetry` are pickier when it comes to validating dependencies and creating lock files and all that, which is part of the issue I am seeing.

Perhaps there is an alternative solution that I am not thinking of either!

**Comment by kclowes** [comment]
Ah, interesting. Thanks for pointing this out! You're seeing this when installing with poetry? Without thinking about it too hard, I think this should be fine to update in v5, especially since we're also using `typing.NoReturn` (which as you pointed out is why `ipfshttpclient` doesn't support `3.7` and `3.7.1`).

**Comment by unparalleled-js** [comment]
> You're seeing this when installing with poetry?

A simple way to reproduce is to create a new project with just a `pyproject.toml` in it with the following contents:

```yaml
[tool.poetry]
name = "my_web3_projet"
authors = ["jules"]
version = "0.1.0"
description = "hi!"

[tool.poetry.dependencies]
python = ">=3.7,<3.10"
web3 = "5.27.0"
```

Notice we are selecting the same python versions that `web.py` has in its `setup.py` file.
When you run `poetry lock` on this project, you get the following error:

```
The current project's Python requirement (>=3.7,<3.10) is not compatible with some of the required packages Python requirement:
    - ipfshttpclient requires Python >=3.6.2,!=3.7.0,!=3.7.1, so it will not be satisfied for Python 3.7 || 3.7.1
  
  Because web3 (5.27.0) depends on ipfshttpclient (0.8.0a2) which requires Python >=3.6.2,!=3.7.0,!=3.7.1, web3 is forbidden.
  So, because my-web3-projet depends on web3 (5.27.0), version solving failed.

  at ~/.poetry/lib/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For ipfshttpclient, a possible solution would be to set the `python` property to ">3.7.0,<3.7.1 || >3.7.1,<3.10"
```

^ Which is how I realized this.

But, thankfully, downstream users are not necessarily blocked by this issue so long as they adjust their version to be `>=3.7.2`.

**Comment by kclowes** [comment]
Can you work around it by changing this: 
```
[tool.poetry.dependencies]
python = ">=3.7,<3.10"
web3 = "5.27.0"
```
to be: 
```
[tool.poetry.dependencies]
python = ">=3.7.2,<3.10"
web3 = ">=5.28.0,<6"
```
?

Is that what you mean by this sentence? 
> But, thankfully, downstream users are not necessarily blocked by this issue so long as they adjust their version to be >=3.7.2.

**Comment by unparalleled-js** [comment]
Yes, exactly! So this is not a blocking issue for me other projects, it was just a bit confusion based on looking the versions I thought I could support

**Comment by kclowes** [comment]
Closed with #2428. Will be released in web3.py v6b3.
